core/commonMain/src/kotlinx/serialization/encoding/ReorderingCompositeEncoder.kt

```diff
@@ -0,0 +1,240 @@
+        }
+        classEncoder.endStructure(descriptor)
+    }
+
+    override fun encodeBooleanElement(descriptor: SerialDescriptor, index: Int, value: Boolean) {
+        bufferEncoding(descriptor, index) { encodeBooleanElement(descriptor, index, value) }
+    }
+
+    override fun encodeByteElement(descriptor: SerialDescriptor, index: Int, value: Byte) {
+        bufferEncoding(descriptor, index) { encodeByteElement(descriptor, index, value) }
+    }
+
+    override fun encodeCharElement(descriptor: SerialDescriptor, index: Int, value: Char) {
+        bufferEncoding(descriptor, index) { encodeCharElement(descriptor, index, value) }
+    }
+
+    override fun encodeDoubleElement(descriptor: SerialDescriptor, index: Int, value: Double) {
+        bufferEncoding(descriptor, index) { encodeDoubleElement(descriptor, index, value) }
+    }
+
+    override fun encodeFloatElement(descriptor: SerialDescriptor, index: Int, value: Float) {
+        bufferEncoding(descriptor, index) { encodeFloatElement(descriptor, index, value) }
+    }
+
+    override fun encodeIntElement(descriptor: SerialDescriptor, index: Int, value: Int) {
+        bufferEncoding(descriptor, index) { encodeIntElement(descriptor, index, value) }
+    }
+
+    override fun encodeLongElement(descriptor: SerialDescriptor, index: Int, value: Long) {
+        bufferEncoding(descriptor, index) { encodeLongElement(descriptor, index, value) }
+    }
+
+    override fun <T : Any> encodeNullableSerializableElement(
+        descriptor: SerialDescriptor,
+        index: Int,
+        serializer: SerializationStrategy<T>,
+        value: T?
+    ) {
+        bufferEncoding(descriptor, index) { encodeNullableSerializableElement(descriptor, index, serializer, value) }
+    }
+
+    override fun <T> encodeSerializableElement(
+        descriptor: SerialDescriptor,
+        index: Int,
+        serializer: SerializationStrategy<T>,
+        value: T
+    ) {
+        bufferEncoding(descriptor, index) { encodeSerializableElement(descriptor, index, serializer, value) }
+    }
+
+    override fun encodeShortElement(descriptor: SerialDescriptor, index: Int, value: Short) {
+        bufferEncoding(descriptor, index) { encodeShortElement(descriptor, index, value) }
+    }
+
+    override fun encodeStringElement(descriptor: SerialDescriptor, index: Int, value: String) {
+        bufferEncoding(descriptor, index) { encodeStringElement(descriptor, index, value) }
+    }
+
+    override fun encodeInlineElement(descriptor: SerialDescriptor, index: Int): Encoder {
+        return BufferingInlineEncoder(descriptor, index)
+    }
+
+    override fun shouldEncodeElementDefault(descriptor: SerialDescriptor, index: Int): Boolean {
+        callShouldEncodeElementDefault = true
+        return true
+    }
+
+    private inner class BufferingInlineEncoder(
+        private val descriptor: SerialDescriptor,
+        private val elementIndex: Int,
+    ) : Encoder {
+        private var nullable = false
```

Comment:
This name is confusing. It is a state that indicates `encodeNotNullMark` was called and needs to be replayed.

Reply:
renamed to `encodeNotNullMarkCalled`